qiskit/result/utils.py

```diff
@@ -56,6 +62,15 @@ def marginal_counts(result, indices=None, inplace=False, format_marginal=False):
             experiment_result.header.memory_slots = len(indices)
             csize = experiment_result.header.creg_sizes
             experiment_result.header.creg_sizes = _adjust_creg_sizes(csize, indices)
+
+            if getattr(experiment_result.data, "memory", None) is not None:
```

Comment:
I think it might be a good idea to add a flag around this to make this an opt-in feature. This is potentially an unexpected behavior change from earlier releases especially with `inplace=True` so having a kwarg like `marginalize_memory=False` and checking that `marginalize_memory == True` here would not surprise anyone.